lnrpc/routerrpc/router_server.go

```diff
@@ -360,13 +360,37 @@ func (s *Server) SendPaymentV2(req *SendPaymentRequest,
 		return err
 	}
 
+	// The payment context is influenced by two user-provided parameters,
+	// the cancelable flag and the payment attempt timeout.
+	// If the payment is cancelable, we will use the stream context as the
+	// payment context. That way, if the user ends the stream, the payment
+	// loop will be canceled.
+	ctx := context.Background()
+	if req.Cancelable {
+		ctx = stream.Context()
+	}
+
+	// The second context parameter is the timeout. If the user provides a
+	// timeout, we will additionally wrap the context in a deadline. If the
+	// user provided 'cancelable' and ends the stream before the timeout is
+	// reached the payment will be canceled.
+	cancel := func() {}
+	if payment.PayAttemptTimeout > 0 {
+		timeout := time.Now().Add(payment.PayAttemptTimeout)
+		ctx, cancel = context.WithDeadline(ctx, timeout)
+	}
+
 	// Send the payment asynchronously.
-	s.cfg.Router.SendPaymentAsync(payment, paySession, shardTracker)
+	s.cfg.Router.SendPaymentAsync(ctx, payment, paySession, shardTracker)
 
 	// Track the payment and return.
-	return s.trackPayment(
-		sub, payHash, stream, req.NoInflightUpdates,
-	)
+	err = s.trackPayment(sub, payHash, stream, req.NoInflightUpdates)
+
+	// Cancel the timeout context. If the context already timed out or if
+	// there was no timeout provided, this will be a no-op.
+	cancel()
```

Comment:
this doesn't cancel the payment directly because `trackPayment` is blocking?

Reply:
This call is to avoid a linter issue that would arise if I wrote above `ctx, _  = context.WithDeadline(ctx, timeout)`. And if I assign the cancel I have to call it. The right place seems to be after `trackPayment` because then the context can be canceled.

Reply:
an alternative could be to place it after `sendPayment` in `SendPaymentAsync`, but not sure that's a good pattern

Reply:
could also silence the linter here with a comment as to why we don't need the context in this case.

Reply:
Maybe @yyforyongyu could opine here as this concerns his suggestion to remove the timeout channel in favor of this context?

Reply:
If we call cancel here I think it will immediately invoke `<-ctx.Done` in `checkContext` since it's async sending. I think we can instead decide the right context to use here (`stream.Context` or background ctx), pass it to `SendPaymentAsync -> sendPayment`, and inside `sendPayment` we create the timeout ctx, and call the `defer cancel()` there.

Reply:
This timeout thing may also be moved to the payment session, but that's a bigger refactor so future PRs.

Reply:
> If we call cancel here I think it will immediately invoke `<-ctx.Done` in `checkContext` since it's async sending.

Before `cancel()` is called there is the `s.trackPayment` call which is blocking. So I think `cancel()` is only called once the track payment call ends which implies that also the async payment loop ended. I think the cancel would then be correct here. We could add clarifying docs here. 

Could also create the timeoutCtx in `sendPayment`. instead.

Reply:
> Before cancel() is called there is the s.trackPayment call which is blocking.

Gotya - yea to me it feels "safer" to do this in `sendPayment`, so any future changes in `s.trackPayment` won't affect the behavior of sending payments.

Reply:
Okay I will change this.